src/main/java/com/salesforce/dynamodbv2/mt/mappers/sharedtable/impl/RandomPartitioningConditionMapper.java

```diff
@@ -30,143 +38,272 @@
+                    }
+
+                    // possibly fan out to assigning multiple fields for SET
+                    String physicalClause = String.join(", ", equalsClauses);
+                    expression = expression.replaceAll(
+                        virtualFieldPlaceholder + " *= *" + virtualValuePlaceholder + "\\b", physicalClause);
                 }
-                request.putExpressionAttributeName(keyFieldName.get(), fieldMapping.getTarget().getName());
             }
         }
+        return expression;
+    }
+
+    private void validateNotUpdatingTablePrimaryKeyFields(String updateExpression, RequestWrapper request) {
+        PrimaryKey primaryKey = tableMapping.getVirtualTable().getPrimaryKey();
+        Set<String> tablePrimaryKeyFields = new HashSet<>();
+        tablePrimaryKeyFields.add(primaryKey.getHashKey());
+        primaryKey.getRangeKey().ifPresent(tablePrimaryKeyFields::add);
+
+        Pattern p = Pattern.compile("#[a-zA-Z0-9]+");
+        Matcher m = p.matcher(updateExpression);
+        while (m.find()) {
+            String placeholder = m.group();
+            String field = request.getExpressionAttributeNames().get(placeholder);
+            if (field != null && tablePrimaryKeyFields.contains(field)) {
+                // not the best error message but this is what local dynamo does?
+                throw new AmazonServiceException("This attribute is part of the key");
+            }
+        }
+    }
+
+    private Set<String> getPlaceholdersForVirtualField(String virtualField, RequestWrapper request) {
+        return Optional.ofNullable(request.getExpressionAttributeNames())
+            .orElse(Collections.emptyMap())
+            .entrySet().stream()
+            .filter(entry -> entry.getValue().equals(virtualField) && !entry.getKey().equals(NAME_PLACEHOLDER))
+            .map(Entry::getKey)
+            .collect(Collectors.toSet());
     }
 
     /**
-     * Extracts literals referenced in expressions and turns them into references to expression names and values.
+     * Extracts literals referenced in an expression and turns them into references to expression names and values.
      *
      * <p>Comments show expected variable values with a sample set of inputs.
      */
-    private String convertFieldNameLiteralsToExpressionNamesInternal(
-        Collection<FieldMapping> fieldMappings, // source = "field", target="field"
-        String conditionExpression, // "field = :value and field2 = :value2 and field = :value3"
-        RequestWrapper request) {
-        String newConditionExpression = conditionExpression;
-        if (conditionExpression != null) {
-            AtomicInteger counter = new AtomicInteger(1);
-            for (FieldMapping fieldMapping : fieldMappings) {
-                String virtualFieldName = fieldMapping.getSource().getName(); // "field"
-                String toFind = " " + virtualFieldName + " ="; // " field ="
```

Comment:
This wasn't working if the virtual field literal is at the beginning of the expression without a space before it